Synonyms APIs - Synonyms Sets documentation (#98015)
elastic/elasticsearch: docs/reference/synonyms/apis/synonyms-apis.asciidoc

[[synonym-management-apis]]

> == Synonym Management APIs

Reviewer: ```suggestion
== Synonyms APIs
```
Plural to match API path. Removed "management" for consistency with other CRUD APIs ("management" is implied). This change would cascade through to a few other places that I didn't mark.

Author: Changing that, thanks Chris!

beta::[]

++++
<titleabbrev>Synonyms Management APIs</titleabbrev>
++++

---

> Synonyms Management APIs let you create and manage synonyms sets.

Reviewer 2: May be we should explain what is a synonyms set with some introduction, with something like this:

"In the past there were only two ways to manage synonyms: providing synonyms inline while defining an analyzer or providing synonyms in a file.  Both these methods have their drawbacks, the former could lead to the out-blown mappings, while the later requires keep synonym files consistent across all nodes.

Synonym Management API is a new way to define and manage synonyms through an API where synonyms will be stored in an internal system index.  A number of synonyms can be grouped in a "synonyms set", and you can have as many sets as needed"....

Author: I like this, but I think this should be part of a blog post explaining the past situation vs the present one. I'm not sure about including the reasons for an API to be created, and talking about "new" might get stale in a few releases from now...

Reviewer 3: I think we can still extract the key points from @mayya-sharipova's suggestion, without "time stamping" the content. For example, we can still outline the relative merits of this new API approach, like how it avoids problems from existing approaches. :)

Reviewer 4: I like the idea of providing a value statement here, and then saving the full in-depth description for a blog post. 👍

Author: Thanks for the feedback! How about this? It's a small rephrasing from Mayya's statement that provides the value proposition and the comparison to other methods. LMKWy'allT!

> Synonym Management API provides a way to define and manage synonyms through an API where synonyms are stored in an internal system index. Related synonyms can be grouped in a "synonyms set". As many synonyms sets can be created as you need for managing synonyms.
> 
> This provides an alternative to the other synonyms definition methods:
> - Inline synonyms in an analyzer definition. This can lead to out-blown mappings.
> - Synonyms file. It requires to keep synonyms files consistent across all nodes.

Reviewer 4: WDYT about these tweaks?

> The synonyms management API provides a convenient way to define and manage synonyms in an internal system index. Related synonyms can be grouped in a "synonyms set". As many synonyms sets can be created as you need for managing synonyms.
> 
> This provides an alternative to the other synonyms definition methods:
> 
> Defining inline synonyms in an analyzer definition, which can lead to [mappings explosion](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-explosion.html) and impact performance 
> Manually uploading a synonyms file and managing consistency across all nodes

On a side note we should probably also update [the synonym token filter doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html) to reference the API as well

Author: Nice! I've put all this in the blender and updated the PR. Thanks for your suggestions!

> On a side note we should probably also update [the synonym token filter doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html) to reference the API as well

For sure - that's something I want to tackle separately, as I'd like to check how to reorg the different synonyms definition methods, potentially as a separate guide. I'll ask for the doc teams help on this 👍

Synonyms sets can be used to configure <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>>.

> Once a synonyms set is used in a synonym token filter, these token filters can be used as part of a <<search-analyzer,search analyzers>>.

Reviewer 2: can be used => must be used or "can only be used"

Synonym token filter can be used  only in search analyzers ...

Author: I'm rephrasing this as "are used as part", to leave the "can" out 👍

You can dynamically change the synonyms used at search time using these APIs to update the synonyms sets.
Your search results will immediately reflect the synonyms set change.

The following APIs can be used to manage synonyms sets:

> * <<put-synoyms-set>>

Reviewer 2: <<put-synoyms-set>> => <<put-synonyms-set>> and others

Author: thanks!

[[synonyms-apis]]
== Synonyms APIs

beta::[]

++++
<titleabbrev>Synonyms APIs</titleabbrev>
++++

---

The synonyms management API provides a convenient way to define and manage synonyms in an internal system index. Related synonyms can be grouped in a "synonyms set".
Create as many synonym sets as you need.

This provides an alternative to:

> - Defining inline synonyms in an analyzer definition, which impacts mapping size and can lead to performance issues.

Reviewer 5: ```suggestion

- Defining inline synonyms in an analyzer definition, which impacts mapping size and can lead to performance issues.
```

Reviewer 5: Without this newline the list doesn't render correctly.

Author: Good catch, thanks! 👍

- Using synonyms files, which implies uploading and managing file consistency on all cluster nodes.

Synonyms sets can be used to configure <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>>.

> Once a synonyms set is used in a synonym token filter, these token filters are used as part of a <<search-analyzer,search analyzers>>.

Reviewer 3: ```suggestion
When synonyms sets are used in synonym token filters, filters are applied as part of the <<search-analyzer,search analyzer>> analysis process.
```

Just trying to make this clearer, not sure if it helps though :)

> You can dynamically change the synonyms used at search time using these APIs to update the synonyms sets.

Reviewer 5: ```suggestion
You can use these APIs to dynamically update synonyms sets used at search time.
```

Reviewer 5: Just a suggestion. The original sentence is hard to read.

> Your search results will immediately reflect the synonyms set change.

Reviewer 3: ```suggestion
Your search results will immediately reflect the synonyms set changes.
```

Plurals get confusing here :)

> The following APIs can be used to manage synonyms sets:

Reviewer 3: ```suggestion
Use the following APIs to manage synonyms sets:
```
